### PR TITLE
Minor refactoring around RequestSpan

### PR DESCRIPTION
**Pull request by hlinnaka**
See commit messages for details.

### CODE REVIEW

File: libs/utils/src/http/endpoint.rs [hunk 1]
```diff
@@ -54,60 +58,58 @@ struct RequestId(String);
...
+    async move {
+        let cancellation_guard = RequestCancelled::warn_when_dropped_without_responding();
+        if log_quietly {
+            debug!("Handling request");
+        } else {
+            info!("Handling request");
+        }
 
-            // Note that we reuse `error::handler` here and not returning and error at all,
-            // yet cannot use `!` directly in the method signature due to `routerify::RouterBuilder` limitation.
```
Comment (hlinnaka) [inline libs/utils/src/http/endpoint.rs:87]: I didn't understand this comment. What does using `!` mean here? Did it mean `?`, perhaps?

I removed this, but if someone understands what it meant or if we should keep it, I'm all ears.

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:87]: I think because we were unable to use infallible as the err with routerify... I am not sure about **not using `error::handler`**, feels like it's usage was retained on purpose.

what's the upside of not using it?

Comment (hlinnaka) [reply libs/utils/src/http/endpoint.rs:87]: > Or I think the idea here was to use is Result<(), !> if I recall correctly.

A-ha, I see. That makes sense, thanks!

Comment (hlinnaka) [reply libs/utils/src/http/endpoint.rs:87]: I put back a comment to explain that, using words that I understand :-).

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:87]: Earlier thread: https://github.com/neondatabase/neon/pull/3755#discussion_r1130573353

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:87]: My understanding of the earlier thread was that we wanted to continue using it regardless of the possible panic, but didn't really record exploring anything about tightening the error to be `ApiError` always. So I understand it was missed.

It also does not seem plausible that there would had been a handler that would had needed to return a different kind of error, otherwise we would had added it to `error::handler`.

Comment (koivunej) [reply libs/utils/src/http/endpoint.rs:87]: Oki, and now I understood on https://github.com/neondatabase/neon/pull/4337#pullrequestreview-1442442050 why `error::handler` is used, it is to be consistent about logging the responses.

While it seems uninteresting it might still make sense to forward just the right lines 105..=112 to a new function, which would be called by this new `request_span` and `error::handler` to log and produce the body.

Comment (hlinnaka) [reply libs/utils/src/http/endpoint.rs:87]: Refactored it that way. I also made `error::handler` a little more lenient, so that it doesn't panic on errors other than ApiError.

File: libs/utils/src/http/error.rs [hunk 2]
```diff
@@ -28,7 +28,7 @@ pub enum ApiError {
 }
 
 impl ApiError {
-    pub fn into_response(self) -> Response<Body> {
```
Comment (LizardWizzard) [inline libs/utils/src/http/error.rs:31]: NIT: Can you elaborate why this change is needed? Previous signature captured the intent of conversion to response by consuming the error. Methods that convert shared references to some other type are usually called `as_foo`, this way it captures the idea that they can be called multiple times on the same object, which as I see it is not needed here

**Review by LizardWizzard — APPROVED** [review]
LGTM, nice, thanks!

One minor nit (re: naming)

File: libs/utils/src/http/error.rs (hunk 2 above)
Comment (koivunej) [reply libs/utils/src/http/error.rs:31]: I don't understand this change either.

Comment (hlinnaka) [reply libs/utils/src/http/error.rs]: I was having use-after-move problems in route_error_handler(), and I fixed / worked around it by using a reference instead. But I figured it out now (Error::downcast() returns the original error in the Err() result, if the downcast fails. I didn't realize that at first). So reverted this back to into_response.
